First need to bring in the numpy library, and get our arrays.
Get numpy library --> import numpy as np.
Copy and Paste arrays a and b from the homework instructions.

In [2]:
import numpy as np
a = np.array([1, 2, 3, 2, 3, 4, 3, 4, 5, 6])
b = np.array([7, 2, 10, 2, 7, 4, 9, 4, 9, 8])

Qestion 1: 
Write a Python/NumPy code block that finds the distinct/unique common items between a and b arrays.
Output should contain only the distinct overlapping values. (For example, if a 2 is found in both array a and array b,
your output should contain only one 2 even if array a contains more than one 2 within it.)

In [10]:
#use np.instersect1d to get distinct/unique common items between two arrays
np.intersect1d(a,b)

array([2, 4])

Question 2: 
Create the following 5x3 array using knowledge you have of Python’s / NumPy’s sequencing functionality
so that you do not need to explicitly key in every integer value.
1 6 11
2 7 12
3 8 13
4 9 14
5 10 15

Do this in a few steps:

1. Create an array with elements 1 to 15
2. Reshape it from a 1D array to an array of 3 x 5
3. Transpose it to get desired order of the 3 x 5 array

In [18]:
#Create array with elements 1 to 15, using np.arange
arr=np.arange(1, 16)
#reshape the array to three rows, five columns 
#name the reshaped array arr1
arr1=arr.reshape(3,5)

In [19]:
print(arr1)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]]


In [20]:
#transpose arr1 
arr1.T

array([[ 1,  6, 11],
       [ 2,  7, 12],
       [ 3,  8, 13],
       [ 4,  9, 14],
       [ 5, 10, 15]])

Question 3: 
Write a Python/Numpy code block that removes from array a1 any items that are also present in array b1.
np.setdiff1d removes items from a1 that are also found in b1

In [21]:
#Copy and Paste arrays a1 and b1 from homework instructions
a1 = np.array([12, 5, 7, 15, 3, 1, 8])
b1 = np.array([14, 6, 3, 11, 19, 12, 5])

In [24]:
#use np.setdiff1d for a1 and b1. a1 will come back without the values that also appear in b1
np.setdiff1d(a1,b1)

array([ 1,  7,  8, 15])

Question 4:
Suppose you have ridden a bicycle from New York City to Key West, Florida. Your bicycle odometer shows
the total miles you have travelled thus far, which you make a note of each day with paper and pencil. Your first two
entries might be ‘55’ and ‘120’, indicating that you rode your bike 55 miles on day 1 and 65 miles on day 2. Your task is to create a NumPy array wherein you can record the cumulative miles you recorded each day during your trip “by hand”.

Then use your Python skills to show the total miles that you rode each day. Assume the trip took a total of 35 days and included at least 4 non-consecutive days where no cycling was possible due to either weather conditions or personal fatigue/soreness. For the remaining 31 days, you would have covered the entire distance between NYC and Key West.

Complete your work within your Jupyter Notebook for this assignment and be sure to include a writeup explaining your
approach to this problem, including the ways in which you decided to make use of NumPy.

First step is to create an array with 35 spots that corresponds to the 35 days on the road:

In [5]:
import numpy as np
#create array 'miles_per_day' using arange function. (1,36) will give the array a range of 35.
miles_per_day=np.arange(1,36)

Google Maps says the distance from NYC to Key West, Florida via bicycle is 1,519 miles.
1,519 miles/ 31days= 49 miles a day (we are not dividing by 35 days, because according to the given word problem,
the biker takes a break for 4 days due to exhaustion or the weather)

The biker needs to average 49 miles a day to get to his desitination within the given time.
For simplicity, I've given my rider the goal of riding 49 miles every day that she is biking.

In [7]:
#edit our array. Make the biker ride 49 miles each day (except for the 4 days that he is not biking)  
miles_per_day[0:36]=49 
miles_per_day[10]=0 
miles_per_day[15]=0
miles_per_day[20]=0
miles_per_day[30]=0

miles_per_day #new values within our array, showing the miles biked per each day of the journey

array([49, 49, 49, 49, 49, 49, 49, 49, 49, 49,  0, 49, 49, 49, 49,  0, 49,
       49, 49, 49,  0, 49, 49, 49, 49, 49, 49, 49, 49, 49,  0, 49, 49, 49,
       49])

At the end of everyday, the biker will mark the number shown on her odometer for cumulative miles rode. 
We can predict her cumulative mileage via the cumulative sum function below. 

In [12]:
miles_per_day.cumsum(axis=0)

array([  49,   98,  147,  196,  245,  294,  343,  392,  441,  490,  490,
        539,  588,  637,  686,  686,  735,  784,  833,  882,  882,  931,
        980, 1029, 1078, 1127, 1176, 1225, 1274, 1323, 1323, 1372, 1421,
       1470, 1519])

In [9]:
#say the cumulative sum array as cum_miles
cum_miles=miles_per_day.cumsum(axis=0)
print(cum_miles)

[  49   98  147  196  245  294  343  392  441  490  490  539  588  637
  686  686  735  784  833  882  882  931  980 1029 1078 1127 1176 1225
 1274 1323 1323 1372 1421 1470 1519]


Let's say our biker misses her goal of 49 miles by 1 miles on the 3rd day, so that her odometer reads 146 instead of 147. She can put the 147 within her cum_miles array by doing cum_miles[2]=146. This allows her to update her array with accurate mileage.

In [ ]:
Now lets say that at the end of the 14th day, her odemeter reads 634 instead of the expected 637. 
She should first change the 13th index in the array to 634.
Now if she wants to see how many actual miles she did that day, instead of the planned 49, she can do subtraction on
the cum_miles array:

In [13]:
#change 637 to 634
cum_miles[13]=634
#use subtraction on the array
cum_miles[13]-cum_miles[12] #she rode 46 miles on the 14th day instead of the planned 49.

46

Question 5:
Importing a data set containing the NYC Water consumption and preform various mathematical and statistical operations using NumPy’s slicing, methods, and functions.

In [14]:
#importing pandas package
import pandas

# supress scientific notation
np.set_printoptions(suppress = True)

# read the data file from GitHub and convert to a NumPy ndarray object
nyc_water = pandas.read_csv('https://raw.githubusercontent.com/jtopor/DAV-5400/master/Week4/Water_Consumption_In_The_New_York_City.csv').values

# verify you have an ndarray object
type(nyc_water)

numpy.ndarray

In [15]:
#print the array to show the data
print(nyc_water)

[[   1979.  7102100.     1512.      213. ]
 [   1980.  7071639.     1506.      213. ]
 [   1981.  7089241.     1309.      185. ]
 [   1982.  7109105.     1382.      194. ]
 [   1983.  7181224.     1424.      198. ]
 [   1984.  7234514.     1465.      203. ]
 [   1985.  7274054.     1326.      182. ]
 [   1986.  7319246.     1351.      185. ]
 [   1987.  7342476.     1447.      197. ]
 [   1988.  7353719.     1484.      202. ]
 [   1989.  7344175.     1402.      191. ]
 [   1990.  7335650.     1424.      194. ]
 [   1991.  7374501.     1469.      199. ]
 [   1992.  7428944.     1369.      184. ]
 [   1993.  7506166.     1368.5     182. ]
 [   1994.  7570458.     1357.7     179. ]
 [   1995.  7633040.     1325.7     174. ]
 [   1996.  7697812.     1297.9     169. ]
 [   1997.  7773443.     1205.5     155. ]
 [   1998.  7858259.     1219.5     155. ]
 [   1999.  7947660.     1237.2     156. ]
 [   2000.  8008278.     1240.4     155. ]
 [   2001.  8024963.5    1184.      148. ]
 [   2002. 

Will now create 4 arrays for the  above 4 columns, which can help in different problem sets and calculations:

In [16]:
#creating new array that containes only the year values from the original set
year=nyc_water[:,0] # defining new array -- ":" = all rows, "0" = first column
print(year) #print year array

[1979. 1980. 1981. 1982. 1983. 1984. 1985. 1986. 1987. 1988. 1989. 1990.
 1991. 1992. 1993. 1994. 1995. 1996. 1997. 1998. 1999. 2000. 2001. 2002.
 2003. 2004. 2005. 2006. 2007. 2008. 2009. 2010. 2011. 2012. 2013. 2014.
 2015. 2016. 2017.]


In [18]:
#creating new array that containes only the nyc population values from the original set
nyc_pop=nyc_water[:,1] # defining new array -- ":" = all rows, "1" = second column

print(nyc_pop) #print nyc_pop array

[7102100.  7071639.  7089241.  7109105.  7181224.  7234514.  7274054.
 7319246.  7342476.  7353719.  7344175.  7335650.  7374501.  7428944.
 7506166.  7570458.  7633040.  7697812.  7773443.  7858259.  7947660.
 8008278.  8024963.5 8041649.  8058334.5 8075020.  8091705.5 8108391.
 8125076.5 8141762.  8158447.5 8175133.  8284098.  8361179.  8422460.
 8471990.  8516502.  8537673.  8622698. ]


In [19]:
#creating new array that containes only the nyc water consumption values from the original set
nyc_consump=nyc_water[:,2]# defining new array -- ":" = all rows, "2" = third column

print(nyc_consump) #print nyc_consump array

[1512.  1506.  1309.  1382.  1424.  1465.  1326.  1351.  1447.  1484.
 1402.  1424.  1469.  1369.  1368.5 1357.7 1325.7 1297.9 1205.5 1219.5
 1237.2 1240.4 1184.  1135.6 1093.7 1099.5 1138.  1069.  1114.  1098.
 1007.5 1039.  1021.  1009.1 1006.1  996.  1009.  1002.   990.2]


In [20]:
#creating new array that containes only the nyc water consumption per calita values from the original set
capita_consump=nyc_water[:,3]# defining new array -- ":" = all rows, "3" = fourth column

print(capita_consump) #print capita_consump array

[213. 213. 185. 194. 198. 203. 182. 185. 197. 202. 191. 194. 199. 184.
 182. 179. 174. 169. 155. 155. 156. 155. 148. 141. 136. 136. 141. 132.
 137. 135. 123. 127. 123. 121. 119. 118. 118. 117. 115.]


Find the maximum yearly NYC consumption of water in millions of gallons per day.
We'll use the max function on the nyc_consump array to bring back the biggest value:

In [21]:
nyc_consump.max()

1512.0

In [23]:
#This could also have been done by using the max function on the nyc_water array
nyc_water.max(0) #brings back the highest value in each column. 
#If we look at the 3rd column - the consumption column - it is the same as our answer above

array([   2017., 8622698.,    1512.,     213.])

How many calendar years are represented within the data set?
We can get this two ways:
Use the shape function on the year array
Use the shape function on the nyc_water array

In [32]:
year.shape[0] 

39

In [28]:
nyc_water.shape[0]

39

In [ ]:
What is the average NYC water consumption per capita?
Use the mean function on the consumption array

In [33]:
capita_consump.mean() 

160.30769230769232

In [35]:
#can also use the mean function on the nyc_water array and look at the result of the fourth column
nyc_water.mean(0)

array([   1998.        , 7789045.80769231,    1234.20769231,
           160.30769231])

What is the standard deviation of NYC water consumption per capita?
Use the std function:

In [37]:
capita_consump.std() 

31.51894330370624

In [38]:
nyc_water.std(0) 
#look at the fourth column - the per capita consumption column. Same answer as above

array([    11.25462868, 470923.52098422,    172.46702157,     31.5189433 ])

Find the increase/decrease values of the nyc population from year to year.
Use the diff function on the nyc_pop array. Diff function finds the numerical differences between the nyc_pop values


In [39]:
np.diff(nyc_pop)

array([-30461. ,  17602. ,  19864. ,  72119. ,  53290. ,  39540. ,
        45192. ,  23230. ,  11243. ,  -9544. ,  -8525. ,  38851. ,
        54443. ,  77222. ,  64292. ,  62582. ,  64772. ,  75631. ,
        84816. ,  89401. ,  60618. ,  16685.5,  16685.5,  16685.5,
        16685.5,  16685.5,  16685.5,  16685.5,  16685.5,  16685.5,
        16685.5, 108965. ,  77081. ,  61281. ,  49530. ,  44512. ,
        21171. ,  85025. ])